source/val/validate_memory.cpp

```diff
@@ -0,0 +1,599 @@
+        // index: the index must be an OpConstant.
+        if (SpvOpConstant != cur_word_instr->opcode()) {
+          return _.diag(SPV_ERROR_INVALID_ID, cur_word_instr)
+                 << "The <id> passed to " << instr_name
+                 << " to index into a "
+                    "structure must be an OpConstant.";
+        }
+        // Get the index value from the OpConstant (word 3 of OpConstant).
+        // OpConstant could be a signed integer. But it's okay to treat it as
+        // unsigned because a negative constant int would never be seen as
+        // correct as a struct offset, since structs can't have more than 2
+        // billion members.
+        const uint32_t cur_index = cur_word_instr->word(3);
+        // The index points to the struct member we want, therefore, the index
+        // should be less than the number of struct members.
+        const uint32_t num_struct_members =
+            static_cast<uint32_t>(type_pointee->words().size() - 2);
+        if (cur_index >= num_struct_members) {
+          return _.diag(SPV_ERROR_INVALID_ID, cur_word_instr)
+                 << "Index is out of bounds: " << instr_name
+                 << " can not find index " << cur_index
+                 << " into the structure <id> '"
+                 << _.getIdName(type_pointee->id()) << "'. This structure has "
+                 << num_struct_members << " members. Largest valid index is "
+                 << num_struct_members - 1 << ".";
+        }
+        // Struct members IDs start at word 2 of OpTypeStruct.
+        auto structMemberId = type_pointee->word(cur_index + 2);
+        type_pointee = _.FindDef(structMemberId);
+        break;
+      }
+      default: {
+        // Give an error. reached non-composite type while indexes still remain.
+        return _.diag(SPV_ERROR_INVALID_ID, cur_word_instr)
+               << instr_name
+               << " reached non-composite type while indexes "
+                  "still remain to be traversed.";
+      }
+    }
+  }
+  // At this point, we have fully walked down from the base using the indeces.
+  // The type being pointed to should be the same as the result type.
+  if (type_pointee->id() != result_type_pointee->id()) {
+    return _.diag(SPV_ERROR_INVALID_ID, &inst)
+           << instr_name << " result type (Op"
+           << spvOpcodeString(static_cast<SpvOp>(result_type_pointee->opcode()))
+           << ") does not match the type that results from indexing into the "
+              "base "
+              "<id> (Op"
+           << spvOpcodeString(static_cast<SpvOp>(type_pointee->opcode()))
+           << ").";
+  }
+
+  return SPV_SUCCESS;
+}
+
+}  // namespace
+
+spv_result_t ValidateMemoryInstructions(ValidationState_t& _) {
+  for (auto& inst : _.ordered_instructions()) {
+    switch (inst.opcode()) {
+      case SpvOpVariable:
+        if (auto error = ValidateVariable(_, inst)) return error;
+        break;
+      case SpvOpImageTexelPointer:
```

Comment:
Move this down to 582 to keep all the empty ones together?

Reply:
done